src/telemetry/writer.cpp

```diff
@@ -0,0 +1,233 @@
+/*
+ * Author: Robertas Norkus
+ * Organisation: HYPED
+ * Date: January 2020
+ * Description:
+ *
+ *    Copyright 2019 HYPED
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <string>
+
+#include "writer.hpp"
+#include "data/data.hpp"
+
+namespace hyped {
+namespace telemetry {
+
+  // additional data points that are displayed in the GUI data section
+  // FEEL FREE TO EDIT. More info: https://github.com/Hyp-ed/hyped-2020/wiki/Adding-new-data-points
+  void Writer::packAdditionalData()
+  {
+    rjwriter_.Key("additional_data");
+    rjwriter_.StartArray();
+
+    // edit below
+
+    // edit above
+
+    rjwriter_.EndArray();
+  }
+
+  // crucial data points that are displayed in various fixed GUI points
+  // NOT EDITABLE
+  void Writer::packCrucialData()
+  {
+    rjwriter_.Key("crucial_data");
+    rjwriter_.StartArray();
+
+    data::Navigation nav_data = data_.getNavigationData();
+    data::StateMachine sm_data = data_.getStateMachineData();
+    add("distance", 0.0, 1250.0, "m", nav_data.distance);
+    add("velocity", 0.0, 250.0, "m/s", nav_data.velocity);
+    add("acceleration", -50.0, 50.0, "m/s^2", nav_data.acceleration);
+    add("status", sm_data.current_state);
+
+    rjwriter_.EndArray();
+  }
+
+  // status data points that are displayed in the GUI status tab
+  // NOT EDITABLE
+  void Writer::packStatusData()
+  {
+    rjwriter_.Key("status_data");
+    rjwriter_.StartArray();
+
+    // TODO(everyone): add all required data points
+
+    rjwriter_.EndArray();
+  }
+
+  Writer::Writer(data::Data& data)
+    : rjwriter_(sb_),
```

Comment:
Be careful about how the members are initialised. In particular, `sb_` is declared after `rjwriter_` in the header file. This means, here `rjwriter_(sb_)` is constructed with uninitialised `sb_` object.